### PR TITLE
[WIP] ✨Disable/enable PortSecurity at port level

### LINKED ISSUE #911: Add a feature to configure Port level PortSecurity
Add a feature to configure Port level PortSecurity such that

- Each port is configured independently. When disabled, both SecurityGroups and allowedAddressPairs values on that port should be removed. specified by `OpenStackMachineTemplate.spec..template.spec.ports[0].disablePortSecurity: false/true`

- When PortSecurity is disabled at the network level, then it is also disabled at each port level. specified by (OpenStackCluster.spec.disablePortSecurity: false)

/kind feature

### PR DESCRIPTION
**Pull request by Xenwar**
This PR adds the capability of enabling/disabling PortSecurity at the port level.

Example Manifest
```yaml
apiVersion: infrastructure.cluster.x-k8s.io/v1alpha4
kind: OpenStackCluster
metadata:
  name: basic-8
  namespace: default
spec:
  disablePortSecurity: false
```

```yaml
apiVersion: infrastructure.cluster.x-k8s.io/v1alpha4
kind: OpenStackMachineTemplate
metadata:
  name: basic-8-control-plane
  namespace: default
spec:
  template:
    spec:
      ports:
      - description: "primary with PortSecurity enabled, required for proper k8s cluster"
        disablePortSecurity: false
        allowedAddressPairs:
        - macAddress: "a4:b8:34:2b:ab:01"
          ipAddress: "10.10.10.10/24"
      - description: "secondary with PortSecurity disabled"
        disablePortSecurity: true
        allowedAddressPairs:
        - macAddress: "a4:b8:34:2b:ab:02"
          ipAddress: "20.20.20.20/24"
```
Example output:
`
```
(openstack) port show primary-port
+-----------------------+--------------------------------------------------------------------+
| Field                 | Value                                                              |
+-----------------------+--------------------------------------------------------------------+
| allowed_address_pairs | ip_address='10.10.10.10/24', mac_address='a4:b8:34:2b:ab:01'       |
| description           | primary with PortSecurity enabled, required for proper k8s cluster |
| port_security_enabled | True                                                               |
| security_group_ids    | 37e42488-1417-4e2c-980f-d2ef71ffbacc                               |
+-----------------------+--------------------------------------------------------------------+
(openstack) port show secondary-port
+-----------------------+--------------------------------------+
| Field                 | Value                                |
+-----------------------+--------------------------------------+
| allowed_address_pairs |                                      |
| description           | secondary with PortSecurity disabled |
| port_security_enabled | False                                |
| security_group_ids    |                                      |
+-----------------------+--------------------------------------+
```
`

Fixes #911
Signed-off-by: Anwar Hassen <anwar.hassen@est.tech>

### CODE REVIEW

File: config/crd/bases/infrastructure.cluster.x-k8s.io_openstackclusters.yaml [hunk 1]
```diff
@@ -1264,6 +1264,11 @@ spec:
                               type: array
                             description:
                               type: string
+                            disablePortSecurity:
+                              description: DisablePortSecurity disables the port security
```
Comment (hidekazuna) [inline config/crd/bases/infrastructure.cluster.x-k8s.io_openstackclusters.yaml:1268]: Where this description come from? Description should be auto generated from a comment in api/v1alpha4/types.go.

File: pkg/cloud/services/compute/instance.go [hunk 2]
```diff
@@ -505,6 +515,28 @@ func (s *Service) getOrCreatePort(eventObject runtime.Object, clusterName string
 	}
 
 	record.Eventf(eventObject, "SuccessfulCreatePort", "Created port %s with id %s", port.Name, port.ID)
+	if portOpts.DisablePortSecurity == true || networkDisabledPortSecurity == true {
```
Comment (hidekazuna) [inline pkg/cloud/services/compute/instance.go:518]: Can't we create port with/without port security not updating after creating port?

File: config/crd/bases/infrastructure.cluster.x-k8s.io_openstackclusters.yaml (hunk 1 above)
Comment (Xenwar) [reply config/crd/bases/infrastructure.cluster.x-k8s.io_openstackclusters.yaml:1268]: You are right. I did not re-generate the yaml files after removing the comment. 
Will add the comment back.

File: pkg/cloud/services/compute/instance.go (hunk 2 above)
Comment (Xenwar) [reply pkg/cloud/services/compute/instance.go:518]: no, we cannot. 
At the port level, PortSecurity is immutable. 
For the network level, there is a bug reporter (https://github.com/kubernetes-sigs/cluster-api-provider-openstack/issues/913)

Comment (hidekazuna) [reply pkg/cloud/services/compute/instance.go:518]: I think it is possible from [gophercloud docs](https://github.com/gophercloud/gophercloud/blob/master/openstack/networking/v2/extensions/portsecurity/doc.go#L94-L120).

Comment (Xenwar) [reply pkg/cloud/services/compute/instance.go:518]: Indeed. The link you provided is more efficient as it does not need retrieving the port to update it, I will update my PR accordingly.

Comment (Xenwar) [reply pkg/cloud/services/compute/instance.go:518]: Thanks a lot

**Comment by hidekazuna** [comment]
@Xenwar I think we need to discuss motivation and spec more. Please see #911 #913

**Comment by Xenwar** [comment]
> @Xenwar I think we need to discuss motivation and spec more. Please see #911 #913

Thanks, I will try it again and update the issue accordingly.

**Comment by Xenwar** [comment]
@hidekazuna 
Having difficult combining these two options. 
The  `ports.Create(s.networkClient,` takes only one of them. Option2 does not allow setting HostID, VNICType and Profile properties. 
Any idea if combining them is possible ?
```

	option1 := portsbinding.CreateOptsExt{
		CreateOptsBuilder: createOpts,
		HostID:            portOpts.HostID,
		VNICType:          portOpts.VNICType,
		Profile:           nil,

	}
	option2 := portsecurity.PortCreateOptsExt{
		CreateOptsBuilder:   createOpts,
		PortSecurityEnabled: &enablePortSecurity,
	}
```

**Comment by hidekazuna** [comment]
> @hidekazuna
> Having difficulty combining these two options.
> The `ports.Create(s.networkClient,` takes only one of them. Option2 does not allow setting HostID, VNICType and Profile properties.
> Any idea if combining them is possible ?
> 
> ```
> 
> 	option1 := portsbinding.CreateOptsExt{
> 		CreateOptsBuilder: createOpts,
> 		HostID:            portOpts.HostID,
> 		VNICType:          portOpts.VNICType,
> 		Profile:           nil,
> 
> 	}
> 	option2 := portsecurity.PortCreateOptsExt{
> 		CreateOptsBuilder:   createOpts,
> 		PortSecurityEnabled: &enablePortSecurity,
> 	}
> ```

@Xenwar  Thanks investigating. I have no idea to combine them as of now.

**Comment by Xenwar** [comment]
@hidekazuna Thanks.

**Comment by iamemilio** [comment]
I just made a similar PR not realizing this was already out there. Lets combine efforts and share homework: https://github.com/kubernetes-sigs/cluster-api-provider-openstack/pull/921

Rather than inheriting the port security settings from the cluster spec, I think its better to just inherit them from the network (which openstack does automatically on create) since these ports can be created on any network in theory. This is why I chose to make it *bool.

**Comment by iamemilio** [comment]
Another thing to watch out for is that because we typically set the security groups at the instance level, OpenStack will apply that security to all port interfaces of that instance. This will cause ugly bugs when setting the port security on those ports.

**Comment by Xenwar** [comment]
> I just made a similar PR not realizing this was already out there. Lets combine efforts and share homework: #921
> 
> Rather than inheriting the port security settings from the cluster spec, I think its better to just inherit them from the network (which openstack does automatically on create) since these ports can be created on any network in theory. This is why I chose to make it *bool.

Thanks, we can combine efforts.
On inheriting portSecurity from the network, my reasoning was there should be a single source of truth for the setting. So, if one changes the setting at the OSCluster spec, then it is propagated to the the Network. 

Also, Network level setting is relevant when PortSecurity is disabled at the network level so the setting is disabled for each port. i.e. Ports do not inherit the setting from the network in all cases.

**Comment by Xenwar** [comment]
> Another thing to watch out for is that because we typically set the security groups at the instance level, OpenStack will apply that security group to all port interfaces of that instance. This will cause ugly bugs when setting the port security on those ports.

We have a use case to enable/disable PortSecurity at the port(s) level. I am not quite sure yet at what point the bug you mentioned could show up, during Port creation or during an update as we have the option of specifying security groups at the port level.

We could probably open a discussion on slack on the this and the other comment.

**Comment by iamemilio** [comment]
/assign @iamemilio 
going to assign myself to this so that I get notifications :)

**Comment by hidekazuna** [comment]
@Xenwar This PR was replaced with #921 and we can close this PR, right?

**Comment by Xenwar** [comment]
> @Xenwar This PR was replaced with #921 and we can close this PR, right?
Thanks for updating the issue. 

I was testing #921  this today and it does not seem to be ready for me.
We should probably unhold this one and proceed with it since it does provide the functionality. 
Probably, a follow up PR can handle the improvements and optimisation. 

@iamemilio can provide more info on this.

**Comment by hidekazuna** [comment]
@Xenwar hold/unfold does not mean ready for review. If PR is not ready for review, rename title starting with WIP, please.

File: docs/book/src/clusteropenstack/configuration.md [hunk 3]
```diff
@@ -327,6 +327,13 @@ A floating IP is created and associated to the bastion host automatically, but y
 
 If `managedSecurityGroups: true`, security group rule opening 22/tcp is added to security groups for bastion, controller, and worker nodes respectively. Otherwise, you have to add `securityGroups` to the `bastion` in `OpenStackCluster` spec and `OpenStackMachineTemplate` spec template respectively.
 
+If PortSecurity disable for all ports either by setting `OpenStackCluster.spec.disablePortSecurity: true` or for each port individually, `managedSecurityGroups: false` is the only valid option. The following is not a valid coonfiguration and is not supported.
```
Comment (hidekazuna) [inline docs/book/src/clusteropenstack/configuration.md:330]: This is in "Accessing nodes through the bastion host via SSH" section, not security group section. How about moving to [Ports](https://github.com/kubernetes-sigs/cluster-api-provider-openstack/blob/master/docs/book/src/clusteropenstack/configuration.md#ports)?

Comment (Xenwar) [reply docs/book/src/clusteropenstack/configuration.md]: Thanks, moved

**Comment by Xenwar** [comment]
This PR is no longer needed. 
Feature is merged by https://github.com/kubernetes-sigs/cluster-api-provider-openstack/pull/921

**Pull request closed without merging**